cognite/client/_api/datapoints.py

```diff
@@ -856,3 +890,125 @@ def _get_datapoints(
         expected_fields = [a for a in aggs] if aggs is not None else ["value"]
         dps = Datapoints._load(res, expected_fields, cognite_client=self.client._cognite_client)
         return dps
+
+
+class _DFColQuery:
+    def __init__(self, start, end, ts_item, aggregates, granularity):
+        self.start = cognite.client.utils._time.timestamp_to_ms(start)
+        self.end = cognite.client.utils._time.timestamp_to_ms(end)
+        self.ts_item = ts_item
+        self.aggregates = aggregates
+        self.granularity = granularity
+        self.granularity_ms = cognite.client.utils._time.granularity_to_ms(granularity)
+        self.result = None
+        self.exception = None
+
+
+class DataFrameFetcher:
+    START_JOBS_PER_THREAD = 2  # how many jobs per worker thread to aim for at the start
+    SLEEP_WAIT_NEW_JOBS = 0.010  # how long to sleep while waiting for new jobs
+
+    def __init__(self, client):
+        self.np, self.pd = utils._auxiliary.local_import("numpy", "pandas")
+        self.client = client
+        self.jobs = []
+        self.lock = threading.Lock()
+
+    def fetch(self, dps_queries, max_workers=None):
+        self.nworkers = max_workers or self.client._config.max_workers
```

Comment:
self.nworkers and self.idle should be instantiated in the constructor.

Reply:
done